cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go

```diff
@@ -0,0 +1,410 @@
+		Id:     toProviderID(vm.ID),
+		Status: toInstanceStatus(vm.Status),
+	}
+}
+
+func toProviderID(nodeID int) string {
+	return fmt.Sprintf("%s%d", providerIDPrefix, nodeID)
+}
+
+func toInstanceStatus(status hcloud.ServerStatus) *cloudprovider.InstanceStatus {
+	if status == "" {
+		return nil
+	}
+
+	st := &cloudprovider.InstanceStatus{}
+	switch status {
+	case hcloud.ServerStatusInitializing:
+	case hcloud.ServerStatusStarting:
+		st.State = cloudprovider.InstanceCreating
+	case hcloud.ServerStatusRunning:
+		st.State = cloudprovider.InstanceRunning
+	case hcloud.ServerStatusOff:
+	case hcloud.ServerStatusDeleting:
+	case hcloud.ServerStatusStopping:
+		st.State = cloudprovider.InstanceDeleting
+	default:
+		st.ErrorInfo = &cloudprovider.InstanceErrorInfo{
+			ErrorClass:   cloudprovider.OtherErrorClass,
+			ErrorCode:    "no-code-hcloud",
+			ErrorMessage: "error",
+		}
+	}
+
+	return st
+}
+
+func newNodeName(n *hetznerNodeGroup) string {
+	return fmt.Sprintf("%s-%d", n.id, rand.Int63())
+}
+
+func buildNodeGroupLabels(n *hetznerNodeGroup) map[string]string {
+	return map[string]string{
+		apiv1.LabelInstanceType:     n.instanceType,
+		apiv1.LabelZoneRegionStable: n.region,
+		nodeGroupLabel:              n.id,
+	}
+}
+
+func getMachineTypeResourceList(m *hetznerManager, instanceType string) (apiv1.ResourceList, error) {
+	typeInfo, _, err := m.client.ServerType.Get(m.apiCallContext, instanceType)
+	if err != nil || typeInfo == nil {
+		return nil, fmt.Errorf("failed to get machine type %s info error: %v", instanceType, err)
+	}
+
+	return apiv1.ResourceList{
+		// TODO somehow determine the actual pods that will be running
+		apiv1.ResourcePods:    *resource.NewQuantity(defaultPodAmountsLimit, resource.DecimalSI),
+		apiv1.ResourceCPU:     *resource.NewQuantity(int64(typeInfo.Cores), resource.DecimalSI),
+		apiv1.ResourceMemory:  *resource.NewQuantity(int64(typeInfo.Memory*1024*1024*1024), resource.DecimalSI),
+		apiv1.ResourceStorage: *resource.NewQuantity(int64(typeInfo.Disk*1024*1024*1024), resource.DecimalSI),
+	}, nil
+}
+
+func serverTypeAvailable(manager *hetznerManager, instanceType string, region string) (bool, error) {
+	serverType, _, err := manager.client.ServerType.Get(manager.apiCallContext, instanceType)
+	if err != nil {
+		return false, err
+	}
+
+	for _, price := range serverType.Pricings {
+		if price.Location.Name == region {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func createServer(n *hetznerNodeGroup) error {
+	StartAfterCreate := true
+	serverCreateResult, _, err := n.manager.client.Server.Create(n.manager.apiCallContext, hcloud.ServerCreateOpts{
+		Name:             newNodeName(n),
+		UserData:         n.manager.cloudInit,
+		Location:         &hcloud.Location{Name: n.region},
+		ServerType:       &hcloud.ServerType{Name: n.instanceType},
+		Image:            &hcloud.Image{Name: n.manager.image},
+		StartAfterCreate: &StartAfterCreate,
+		Labels: map[string]string{
+			nodeGroupLabel: n.id,
```

Comment:
Same as above: could include a label for the cluster name as well, but that depends on whether a Hetzner Cloud project should have more than one cluster anyway, or if clusters running in the same project can rely on choosing distinct node group names.
It might just overcomplicate the implementation though, so I'd be fine with leaving it the way it is.